# Baseline Benchmarks Analysis

## Load Data & Dependencies

In [1]:
from sklearn.metrics import accuracy_score, classification_report
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from tensorflow.keras import Input
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
from keras.backend import clear_session
from xgboost import XGBClassifier
import numpy as np
import pandas as pd

X = pd.read_csv('../../data/ogbn/processed/ogn_arxiv.csv')
y = pd.read_csv('../../data/ogbn/processed/labels.csv')

FileNotFoundError: [Errno 2] No such file or directory: '../../data/ogbn/processed/ogn_arxiv.csv'

**Dataset Shape:**
- Features (X): (169,343, 131)
- Labels (y): (169,343, 1)

In [16]:
print(X.shape)
print(y.shape)

(169343, 131)
(169343, 1)


## Split Data Into Train, Validation, & Test Sets

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y['0'].values, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.25, random_state=1)

Benchmarks: Multinomial Logistic Regression

## Benchmarks: Extreme Gradient Boosting

The ogbn-arxiv dataset presents significant challenges due to class imbalance and feature overlap:

- **Severe class imbalance**: A few dominant classes have 20,000+ samples each, while many classes have only 1,000-3,000 samples, and some have fewer than 500 samples
- **Feature overlap**: t-SNE visualization shows features create overlapping clusters with no clear separation between classes
- **Missing graph structure**: Without leveraging citation relationships, the model loses crucial contextual information

### XGBoost Implementation

In [18]:
from sklearn.metrics import f1_score

xgb_classifier = XGBClassifier(
    objective='multi:softmax',
    eval_metric='mlogloss',
    n_estimators=200,
    learning_rate=0.1,
    max_depth=8,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=1,
    n_jobs=-1,
    early_stopping_rounds=10
)

xgb_classifier.fit(
    X_train,
    y_train,
    eval_set=[(X_val, y_val)],
    verbose=False
)

val_predictions = xgb_classifier.predict(X_val)

val_accuracy = accuracy_score(y_val, val_predictions)
val_macro_f1 = f1_score(y_val, val_predictions, average='macro')

predictions = xgb_classifier.predict(X_test)

[0]	validation_0-mlogloss:3.25492
[1]	validation_0-mlogloss:3.04821
[2]	validation_0-mlogloss:2.89278
[3]	validation_0-mlogloss:2.76977
[4]	validation_0-mlogloss:2.66710
[5]	validation_0-mlogloss:2.57849
[6]	validation_0-mlogloss:2.50234
[7]	validation_0-mlogloss:2.43509
[8]	validation_0-mlogloss:2.37480
[9]	validation_0-mlogloss:2.32113
[10]	validation_0-mlogloss:2.27252
[11]	validation_0-mlogloss:2.22790
[12]	validation_0-mlogloss:2.18732
[13]	validation_0-mlogloss:2.14981
[14]	validation_0-mlogloss:2.11561
[15]	validation_0-mlogloss:2.08397
[16]	validation_0-mlogloss:2.05481
[17]	validation_0-mlogloss:2.02744
[18]	validation_0-mlogloss:2.00180
[19]	validation_0-mlogloss:1.97782
[20]	validation_0-mlogloss:1.95492
[21]	validation_0-mlogloss:1.93438
[22]	validation_0-mlogloss:1.91503
[23]	validation_0-mlogloss:1.89651
[24]	validation_0-mlogloss:1.87933
[25]	validation_0-mlogloss:1.86309
[26]	validation_0-mlogloss:1.84808
[27]	validation_0-mlogloss:1.83340
[28]	validation_0-mlogloss:1.8

### XGBoost Results

In [1]:
accuracy = accuracy_score(y_test, predictions)
report = classification_report(y_test, predictions, output_dict=True, zero_division=0)

f1 = report['macro avg']['f1-score']
recall = report['macro avg']['recall']
precision = report['macro avg']['precision']

print(f"Validation Accuracy: {val_accuracy:.4f}")
print(f"Validation Macro-F1: {val_macro_f1:.4f}")

print(f"Accuracy: {accuracy:.4f}")
print(f"F1: {f1:.4f}")
print(f"Recall: {recall:.4f}")
print(f"Precision: {precision:.4f}")

NameError: name 'accuracy_score' is not defined

**Performance Analysis:**
- **Accuracy (55.86%)**: Out of all paper classifications across 40 subject areas, 56% were assigned to the correct category. XGBoost still misclassifies nearly half of all papers.
- **Precision (47.00%)**: When XGBoost predicts a paper belongs to a specific subject area, it's correct 47% of the time. This suggests significant confusion between similar subjects.
- **Recall (30.08%)**: XGBoost only identifies 30% of papers that actually belong to each subject area. This means for any given subject area, about 70% of papers that should be classified there are being missed. Some subject areas are likely being severely under-detected.
- **F1 Score (33.80%)**: The harmonic mean of precision and recall across all 40 classes. The low F1 indicates the model struggles significantly with class discrimination, with many subject areas probably performing much worse than others.

## Benchmarks: Small Deep‑Learning Text Classifier (MLP)
### MLP Implementation

In [21]:
clear_session()

num_classes = len(set(y_train))
model = Sequential([
    Input(shape=(X_train.shape[1],)),
    Dense(128, activation='relu', name='dense_1'),
    Dropout(0.3),
    Dense(64, activation='relu', name='dense_2'),
    Dense(num_classes, activation='softmax', name='output')
])

model.compile(
    optimizer= Adam(learning_rate=1e-3),
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy'],
)

model.fit(
    X_train, y_train,
    validation_data=(X_val, y_val),
    epochs=20,
    batch_size=32,
    verbose=True,
)

y_predictions_probability = model.predict(X_test)
nn_predictions = np.argmax(y_predictions_probability, axis=1)

Epoch 1/20
3176/3176 ━━━━━━━━━━━━━━━━━━━━ 12s 3ms/step - accuracy: 0.4211 - loss: 2.1193 - val_accuracy: 0.5261 - val_loss: 1.6404
Epoch 2/20
3176/3176 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.5229 - loss: 1.6362 - val_accuracy: 0.5363 - val_loss: 1.5935
Epoch 3/20
3176/3176 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.5414 - loss: 1.5677 - val_accuracy: 0.5445 - val_loss: 1.5619
Epoch 4/20
3176/3176 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.5495 - loss: 1.5349 - val_accuracy: 0.5492 - val_loss: 1.5431
Epoch 5/20
3176/3176 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.5524 - loss: 1.5067 - val_accuracy: 0.5510 - val_loss: 1.5370
Epoch 6/20
3176/3176 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.5663 - loss: 1.4712 - val_accuracy: 0.5596 - val_loss: 1.5179
Epoch 7/20
3176/3176 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.5659 - loss: 1.4675 - val_accuracy: 0.5564 - val_loss: 1.5236
Epoch 8/20
3176/3176 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.5725 - loss: 1

In [22]:
f1 = report['macro avg']['f1-score']
recall = report['macro avg']['recall']
precision = report['macro avg']['precision']

print(f"Accuracy: {accuracy:.4f}")
print(f"F1: {f1:.4f}")
print(f"Recall: {recall:.4f}")
print(f"Precision: {precision:.4f}")

Accuracy: 0.5441
F1: 0.3245
Recall: 0.2877
Precision: 0.4724


**Performance Analysis:**
- **Accuracy (54.41%)**: Out of all paper classifications across 40 subject areas, 54% were assigned to the correct category. The MLP performs slightly worse than XGBoost, still misclassifying about half of all papers.
- **Precision (47.24%)**: When the MLP predicts a category a given paper belongs to, it's correct 47% of the time. Similar to XGBoost, confusion between similar subjects exists.
- **Recall (28.77%)**: The MLP only identifies 29% of papers that actually belong to each subject area. This means for any given subject area, about 71% of papers that should be classified there are being missed. Some subject areas are likely being severely under-detected.
- **F1 Score (32.45%)**: The harmonic mean of precision and recall across all 40 classes is around 32%. The low F1 indicates the model struggles significantly with class discrimination, with many subject areas probably performing much worse than others.

## Performance Explanation

Both models face fundamental challenges that limit their effectiveness:

1. **Severe Class Imbalance**: The target distribution shows a few dominant classes with 20,000+ samples, most classes with 1,000-2,000 samples, and some with fewer than 500 samples. This introduces bias toward dominant classes.

2. **Feature Overlap**: The t-SNE visualization reveals that node features create overlapping "blobs" with heavy mixing between different classes and no clear clustering or separation. Classes are not linearly separable in the current feature space.

3. **Missing Graph Structure**: The most critical limitation is the inability to preserve and leverage the graph structure. For ogbn-arxiv, citation relationships provide crucial context about paper relationships and subject areas. Without this graph information, the models are essentially performing text classification on paper abstracts/titles alone, missing the most informative signal.

The low recall across both models suggests they are biased toward predicting only major classes, while the moderate precision indicates that when they do predict major classes, they're often correct. However, they rarely predict minority classes, leading to the poor overall F1 scores.